content/5_Plat/Segtree_Ext.mdx

```diff
@@ -111,6 +116,156 @@ int main() {
+should walk left or right. But with sum, we also need to consider the range $[1, l)$ and how it might affect our answer.
+
+To handle keeping track of the prefix of $[1, l)$, we first set our prefix value to a neutral value, then join it with our left child's answer every time we walk right.
+
+**Time Complexity:** $\mathcal{O}(Q\log{Q})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+template <class T> class SumSegmentTree {
+  private:
+	const T DEFAULT = 0;
+
+	int len;
+	vector<T> segtree;
+
+	T combine(const T &a, const T &b) { return a + b; }
+
+	void build(const vector<T> &arr, int at, int at_left, int at_right) {
+		if (at_left == at_right) {
+			segtree[at] = arr[at_left];
+			return;
+		}
+		int mid = (at_left + at_right) / 2;
+		build(arr, 2 * at, at_left, mid);
+		build(arr, 2 * at + 1, mid + 1, at_right);
+		segtree[at] = combine(segtree[2 * at], segtree[2 * at + 1]);
+	}
+
+	void set(int ind, T val, int at, int at_left, int at_right) {
+		if (at_left == at_right) {
+			segtree[at] = val;
+			return;
+		}
+		int mid = (at_left + at_right) / 2;
+		if (ind <= mid) {
+			set(ind, val, 2 * at, at_left, mid);
+		} else {
+			set(ind, val, 2 * at + 1, mid + 1, at_right);
+		}
+		segtree[at] = combine(segtree[2 * at], segtree[2 * at + 1]);
+	}
+
+	T range_sum(int start, int end, int at, int at_left, int at_right) {
+		if (at_right < start || end < at_left) { return DEFAULT; }
+		if (start <= at_left && at_right <= end) { return segtree[at]; }
+		int mid = (at_left + at_right) / 2;
+		T left_res = range_sum(start, end, 2 * at, at_left, mid);
+		T right_res = range_sum(start, end, 2 * at + 1, mid + 1, at_right);
+		return combine(left_res, right_res);
+	}
+
+	int walk(int at, int at_left, int at_right, int max_val, int pref_res) {
+		if (at_left == at_right) { return at_left; }
+		int mid = (at_left + at_right) / 2;
+		int sum_left = pref_res + segtree[2 * at];
+		if (sum_left >= max_val) {
+			return walk(2 * at, at_left, mid, max_val, pref_res);
+		}
+		return walk(2 * at + 1, mid + 1, at_right, max_val, sum_left);
+	}
+
+  public:
+	SumSegmentTree(int len) : len(len) { segtree = vector<T>(len * 4, DEFAULT); };
+
+	SumSegmentTree(const vector<T> &arr) : len(arr.size()) {
+		segtree = vector<T>(len * 4, DEFAULT);
+		build(arr, 1, 0, len - 1);
+	}
+
+	void set(int ind, T val) { set(ind, val, 1, 0, len - 1); }
+
+	T range_sum(int start, int end) { return range_sum(start, end, 1, 0, len - 1); }
+
+	int walk(int val) { return walk(1, 0, len - 1, val, 0); }
+};
+
+int main() {
+	int query_num;
+	cin >> query_num;
+	vector<int> indices;
+	vector<pair<char, int>> updates(query_num);
+	for (int i = 0; i < query_num; i++) {
+		char type;
+		int val;
+		cin >> type >> val;
+		updates[i] = {type, val};
+		indices.push_back(val);
+	}
+
+	sort(begin(indices), end(indices));
+	indices.erase(unique(begin(indices), end(indices)), end(indices));
```

Comment:
wait surely std set works :clueless:

Reply:
wdym? like put everything in a set and then put it into a vector again? 

i prefer to do it this way b/c it's faster & shorter

Reply:
ooops forgor you can't index a set